StorageUtility/Windows/Device.cpp

```diff
@@ -0,0 +1,115 @@
+{
+    Data.insert({ eDeviceDataType::PSP_DEVINFO_DATA, m_DevInfoData });
+    Data.insert({ eDeviceDataType::PSP_DEVICE_INTERFACE_DATA, m_DevInterfaceData });
+    Data.insert({ eDeviceDataType::PSP_INTERFACE_DEVICE_DETAIL_DATA, m_DevDetailData });
+}
+
+DeviceHandle cDevice::Handle()
+{
+    eErrorCode errorCode;
+    DeviceHandle deviceHandle;
+
+    errorCode = GetStorageDeviceHandle(m_DevDetailData->DevicePath, deviceHandle);
+    if (eErrorCode::None != errorCode)
+    {
+        throw std::exception("Failed to create device handle", (int)errorCode);
+    }
+    return deviceHandle;
+}
+
+eErrorCode cDevice::GetStorageDeviceHandle(const String& DevicePath, HANDLE& Handle)
+{
+    Handle =
+        CreateFile(
+        DevicePath.c_str(),
+        GENERIC_READ | GENERIC_WRITE,
+        FILE_SHARE_READ | FILE_SHARE_WRITE,
+        NULL,
+        OPEN_EXISTING,
+        FILE_ATTRIBUTE_NORMAL,
+        NULL
+        );
+
+    if (INVALID_HANDLE_VALUE == Handle)
+    {
+        return(eErrorCode::Io);
+    }
+
+    return(eErrorCode::None);
+}
+
+void cDevice::AllocateMemories(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData)
+{
+    m_DevInfoDataHeap = GetProcessHeap();
+    m_DevInterfaceDataHeap = GetProcessHeap();
+    m_DevDetailDataHeap = GetProcessHeap();
+
+    U32 size = sizeof(DevInfoData->cbSize) + sizeof(DevInfoData->ClassGuid) + sizeof(DevInfoData->DevInst) + sizeof(DevInfoData->Reserved);
+    m_DevInfoData = (PSP_DEVINFO_DATA)HeapAlloc(m_DevInfoDataHeap, HEAP_GENERATE_EXCEPTIONS, size);
+
+    size = sizeof(DevInterfaceData->cbSize) + sizeof(DevInterfaceData->Flags) + sizeof(DevInterfaceData->InterfaceClassGuid) + sizeof(DevInterfaceData->Reserved);
+    m_DevInterfaceData = (PSP_DEVICE_INTERFACE_DATA)HeapAlloc(m_DevInterfaceDataHeap, HEAP_GENERATE_EXCEPTIONS, size);
+
+    m_DevDetailData = (PSP_INTERFACE_DEVICE_DETAIL_DATA)HeapAlloc(m_DevDetailDataHeap, HEAP_GENERATE_EXCEPTIONS, SizeOfDevDetailData);
+}
+
+void cDevice::CopyMemories(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData)
+{
+    m_DevInfoData->cbSize = DevInfoData->cbSize;
+    m_DevInfoData->ClassGuid = DevInfoData->ClassGuid;
+    m_DevInfoData->DevInst = DevInfoData->DevInst;
+    m_DevInfoData->Reserved = DevInfoData->Reserved;
+
+    m_DevInterfaceData->cbSize = DevInterfaceData->cbSize;
+    m_DevInterfaceData->Flags = DevInterfaceData->Flags;
+    m_DevInterfaceData->InterfaceClassGuid = DevInterfaceData->InterfaceClassGuid;
+    m_DevInterfaceData->Reserved = DevInterfaceData->Reserved;
+
+    size_t devicePathLength = (std::wcslen(DevDetailData->DevicePath) + 1);
+    m_DevDetailData->cbSize = DevDetailData->cbSize;
+    wmemcpy(m_DevDetailData->DevicePath, DevDetailData->DevicePath, devicePathLength);
```

Comment:
1. You should be using the more secure version of this function: `memcpy_s` or `wmemcpy_s`
2. Are you sure the content will be wide characters?